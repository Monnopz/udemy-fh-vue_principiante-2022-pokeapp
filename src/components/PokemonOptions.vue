<template>
  <div>
    <div class="options-container">
        <ul>
            <!-- El equipo de Vue agrupo variables, ojetos y metodos con simbolo de dolar $; en este caso $emit -->
            <!-- $emit(nombreMetodo, valoraenviar) -->
            <!-- Como buena practica de parte del equipo de Vue, al emitir usar upperCamelCase (selectionPokemon) y al escuchar el evento usar keba-cas(selection-pokemon) -->
            <!-- Para escuchar un evento simplemente es con arroba (@); @selection-pokemon="metodoDelPadreAEjecutar" -->
            <li v-for="pokemon in pokemons" :key="pokemon.id" @click="$emit('selectionPokemon', pokemon.id)">{{pokemon.name}}</li>
        </ul>
    </div>
  </div>
</template>

<script>
export default {
    name: 'PokemonOptions',
    props: {
        pokemons: {
            type: Array,
            required: true
        }
    }
}
</script>

<style scoped>
/* Pokemon Options */
ul {
    list-style-type: none;
    padding-left: 0;
}
li {
    background-color: white;
    border-radius: 5px;
    border: 1px solid rgba(0, 0, 0, 0.2);
    cursor: pointer;
    margin-bottom: 10px;
    width: 250px;
}

li:hover {
    background-color: rgba(0, 0, 0, 0.05);
}

.options-container {
    display: flex;
    justify-content: center;
}
</style>