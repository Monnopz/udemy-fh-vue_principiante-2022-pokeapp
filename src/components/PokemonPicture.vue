<template>
  <div>
    <div class="pokemon-container">
        <img v-if="!showPokemon" class="hidden-pokemon" :src="imgSrc" alt="pokemon">
        <img v-else class="fade-in" :src="imgSrc" alt="pokemon">
    </div>
  </div>
</template>

<script>
export default {

    name: 'PokemonPicture',
    props: {
        pokemonId: {
            type: Number, 
            required: true
        },
        showPokemon: {
            type: Boolean,
            required: true,
            default: false
        }
    },

    computed: {
        imgSrc() {
            return `https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/${this.pokemonId}.svg`
        }
    }

}
</script>

// scoped es solo para que se aplique a este componente
<style scoped>

/* Pokemon Picture */
.pokemon-container {
    height: 200px;
}
img {
    height: 200px;
    position: absolute;
    /* right: 32%; */
    left: 0;
    right: 0;
    margin: auto;
    /* Con esto se evita el drag n drop */
    user-select: none; 
    -moz-user-select: none;
    -ms-user-select: none;
    -webkit-user-drag: none;
    -webkit-user-select: none;
}
.hidden-pokemon {
    filter: brightness(0);
}
</style>