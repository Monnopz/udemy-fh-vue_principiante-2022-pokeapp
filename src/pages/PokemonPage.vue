<template>
  <h1 v-if="!pokemon">Espere por favor...</h1>
  <div v-else>
    <h1>¿Quién es este pokémon?</h1>
    <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon"></PokemonPicture>
    <!-- El equipo de Vue agrupo variables, ojetos y metodos con simbolo de dolar $; en este caso $emit -->
    <!-- $emit(nombreMetodo, valoraenviar) -->
    <!-- Como buena practica de parte del equipo de Vue, al emitir usar upperCamelCase (selectionPokemon) y al escuchar el evento usar keba-cas(selection-pokemon) -->
    <!-- Esto de cambio de upperCamelCase a kebacase es posible gracias a que Vue lo hace desde un inicio al querer usar nuestros componentes propios con el nombre que les dimos o con keba case -->
    <!-- Para escuchar un evento simplemente es con arroba (@); @selection-pokemon="metodoDelPadreAEjecutar" -->
    <PokemonOptions @selection-pokemon="checkAnswer(1, $event)" :pokemons="pokemonArr"></PokemonOptions>
    <!-- El usar el template, aqui en Vue esta agrupado para manejar la condicion, pero al renderizar el DOM, este agrupador no aparecerá -->
    <template v-if="showAnswer" class="fade-in">
        <h2>{{ message }}</h2>
        <button @click="newGame()">Nuevo Juego</button>
    </template>
  </div>
</template>

<script>
//Este tipo de imports son la manera tradicional y cuando el componente o la aplicacion es construida automaticamente forman parte del bundle
// @/ significa que empezará desde la carpeta src
import PokemonPicture from '@/components/PokemonPicture.vue'
import PokemonOptions from '@/components/PokemonOptions.vue'

import getPokemonOptions from '@/helpers/getPokemonOptions'

//los hooks del ciclo de vida de un componente de Vue son metodos que se llaman cuando se crea, se actualiza o se destruye dicho componente
export default {
    name: 'PokemonPage',
    components: {
        PokemonPicture,
        PokemonOptions
    },
    data() {
        return {
            pokemonArr: [],
            pokemon: null,
            showPokemon: false,
            showAnswer: false,
            message: ''
        }
    },
    methods: {
        async mixPokemonArray() {
            this.pokemonArr = await getPokemonOptions()

            //Math.floor redondea y quita todos los decimales
            //Math.random() * 4 es para que haga del 1 al 4
            const rndInt = Math.floor(Math.random() * 4)

            this.pokemon = this.pokemonArr[rndInt]
        },

        checkAnswer(pokemonId) {
            // Se puede mandar por referencia a llamar el metodo o si no se puede poner con parentesis
            // Para mandar mas atributos, se mandan los atributos y al final lo que emite el evento
            // Por ejemplo, (a, b, c, $event)
            //Aunque en Js se pueden mandar valores en funciones por referencia
            this.showPokemon = true;
            this.showAnswer = true;
            if(pokemonId === this.pokemon.id) {
                this.message = `Correcto, es ${this.pokemon.name}`
                
            }
            else{
                this.message = `Oops, era ${this.pokemon.name}`
            }
        },
        newGame(){
            this.showPokemon = false
            this.showAnswer = false
            this.pokemonArr = []
            this.pokemon = null
            this.message = ''
            this.mixPokemonArray()
        }
    },
    mounted(){
        this.mixPokemonArray()
    }
}
</script>