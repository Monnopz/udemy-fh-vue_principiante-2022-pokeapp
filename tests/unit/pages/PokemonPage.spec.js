import { shallowMount, mount } from '@vue/test-utils'
import PokemonPage from '@/pages/PokemonPage'
import PokemonPicture from '@/components/PokemonPicture'
import PokemonOptions from '@/components/PokemonOptions'

import { mockPokemons } from '../mocks/pokemons.mock'

describe('PokemonPage Component', () => {{

    let wrapper

    beforeEach(() => {
        // mixPokemonArrSpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArray') //Se ejecuts aqui antes de hacer el mount. Si se usa aqui hay que hacer el clean de los mocks (jetSpy.clearAllMocks)
        wrapper = shallowMount(PokemonPage)
    })

    // test('debe de hacer match con el snapshot', () => { Esta prueba no ayuda mucho debido a que este componente puede o no tener pokemon
    //     expect(wrapper.html()).toMatchSnapshot()
    // });

    // const mixPokemonArr= jest.spyOn(PokemonPage.methods,'mixPokemonArr') // para espiar el método dentro del componente original

    // const mixPokemonArr= jest.spyOn(wrapper.vm,'mixPokemonArr') // para espiar el método del objeto virtual

    test('debe de llamar el mixPokemonArray al montar ', () => {

        const mixPokemonArrSpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArray') //O se puede crear aqui y montar otro wrapper
        const wrapper = shallowMount(PokemonPage)
        expect(mixPokemonArrSpy).toHaveBeenCalled()

    });

    test('debe de hacer match con el snapshot cuando cargan los pokemons', () => {
        //El snapshot comun espera y carga informacion una vez (expect(wrapper.html()).toMatchSnapshot())
        //El shallowMount monta de manera sutil y no renderiza internamente porque dispara ciertos porcesos del ciclo de vida de los componentes probados
        //Un stub es un stand in y simula su comportamiento
        //El mount si monta los componentes y los renderiza, aunque este tipo de prueba es mas pesada
        const wrapper = shallowMount(PokemonPage, { //Tambien se puede montar la data
            data() {
                return {
                    pokemonArr: mockPokemons,
                    pokemon: mockPokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })
        expect(wrapper.html()).toMatchSnapshot()
    });

    test('debe de mostrar los componentes de PokemonPicture y PokemonOptions', () => {

        //PokemonPicture deben de existir
        //PokemonOptions deben de existir
        //PokemonPicture attribute pokemonId === 1
        //PokemonOptions attribute pokemons toBe true
        
        // //Mi solucion
        // const wrapper = shallowMount(PokemonPage, { //Se monta la data de PokemonPge para que el html sea cargado ya con los componentes que deben aparecer
        //     data() {
        //         return {
        //             pokemonArr: mockPokemons,
        //             pokemon: mockPokemons[0],
        //             showPokemon: false,
        //             showAnswer: false,
        //             message: ''
        //         }
        //     }
        // })

        // const pokemonPictureComponent = wrapper.find('pokemon-picture-stub') //Encuentra el componente
        // const pokemonOptionsComponent = wrapper.find('pokemon-options-stub') //Encuentra el componente

        // expect(pokemonPictureComponent.exists()).toBeTruthy() //Realiza las pruebas de si existe el componente
        // expect(pokemonOptionsComponent.exists()).toBeTruthy() //Realiza las pruebas de si existe el componente

        // expect(pokemonPictureComponent.attributes('pokemonid')).toBe("1") //Verifica que el pokemonid sea 1
        // expect(pokemonOptionsComponent.attributes('pokemons')).not.toBe('') //Verifica que el array de pokemons si esté (que no este vacio)

        

        //Solucion Fernando

        const wrapper = shallowMount(PokemonPage, { //Se monta la data de PokemonPge para que el html sea cargado ya con los componentes que deben aparecer
            data() {
                return {
                    pokemonArr: mockPokemons,
                    pokemon: mockPokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        expect(wrapper.find('pokemon-picture-stub').exists()).toBeTruthy()
        expect(wrapper.find('pokemon-options-stub').exists()).toBeTruthy()

        expect(wrapper.find('pokemon-picture-stub').attributes('pokemonid')).toBe("1")
        expect(wrapper.find('pokemon-options-stub').attributes('pokemons')).toBeTruthy()

        //Otra forma de hacerlo
        // const wrapper = shallowMount(PokemonPage, {
        //     data() {
        //         return {
        //             pokemonArr: mockPokemons,
        //             pokemon: mockPokemons[0],
        //             showPokemon: false,
        //             showAnswer: false,
        //             message: ""
        //         }
        //     }
        // })
        // const pokemonPictureComponent = wrapper.findComponent(PokemonPicture)
        // const pokemonOptionsComponent = wrapper.findComponent(PokemonOptions)
        // const { pokemonId } = pokemonPictureComponent.props()
        // const { pokemons: arrPokemon } = pokemonOptionsComponent.props()
 
        // expect(pokemonPictureComponent.exists()).toBeTruthy()
        // expect(pokemonOptionsComponent.exists()).toBeTruthy()
        // expect(pokemonId).toBe(1)
        // expect(arrPokemon).toStrictEqual(mockPokemons)

    });

    test('pruebas con checkAnswer', async () => {

        //wrapper.find(); encontrar algo en el componente; pero utilizar otros metodos del componente
        //wrapper.vm.showPokemon; utilizar una prop del componente montado
        
        const wrapper = shallowMount(PokemonPage, { //Se monta la data de PokemonPge para que el html sea cargado ya con los componentes que deben aparecer
            data() {
                return {
                    pokemonArr: mockPokemons,
                    pokemon: mockPokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        await wrapper.vm.checkAnswer(1) //Un await para cuando hay modificaciones en las propiedades reactivas

        expect(wrapper.find('h2').exists()).toBeTruthy()
        expect(wrapper.vm.showPokemon).toBeTruthy()
        expect(wrapper.find('h2').text()).toBe(`Correcto, es ${mockPokemons[0].name}`)

        await wrapper.vm.checkAnswer(10)

        expect(wrapper.vm.message).toBe(`Oops, era ${mockPokemons[0].name}`)

    });

}})