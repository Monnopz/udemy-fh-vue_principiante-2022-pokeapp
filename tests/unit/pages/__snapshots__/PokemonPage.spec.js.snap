// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Component debe de hacer match con el snapshot cuando cargan los pokemons 1`] = `
<div>
  <h1>¿Quién es este pokémon?</h1>
  <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub><!-- El equipo de Vue agrupo variables, ojetos y metodos con simbolo de dolar $; en este caso $emit -->
  <!-- $emit(nombreMetodo, valoraenviar) -->
  <!-- Como buena practica de parte del equipo de Vue, al emitir usar upperCamelCase (selectionPokemon) y al escuchar el evento usar keba-cas(selection-pokemon) -->
  <!-- Esto de cambio de upperCamelCase a kebacase es posible gracias a que Vue lo hace desde un inicio al querer usar nuestros componentes propios con el nombre que les dimos o con keba case -->
  <!-- Para escuchar un evento simplemente es con arroba (@); @selection-pokemon="metodoDelPadreAEjecutar" -->
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub><!-- El usar el template, aqui en Vue esta agrupado para manejar la condicion, pero al renderizar el DOM, este agrupador no aparecerá -->
  <!--v-if-->
</div>
`;
